Remove fomantic button module

silverwind opened the pull request:
CSS-only module. Button colors are reduced to this:

<img width="639" alt="Screenshot 2024-04-14 at 15 36 07" src="https://github.com/go-gitea/gitea/assets/115237/882d6c02-d1de-44f2-b707-db02a9f5070d">

silverwind commented on templates/repo/issue/view_content/sidebar.tmpl:
```diff
@@ -572,7 +572,7 @@
 		{{end}}
 
-		<button class="tw-mt-1 fluid ui show-modal button {{if .Issue.IsLocked}} negative {{end}}" data-modal="#lock">
+		<button class="tw-mt-1 fluid ui show-modal button{{if .Issue.IsLocked}} negative{{end}}" data-modal="#lock">
```
```suggestion
		<button class="tw-mt-1 fluid ui show-modal button{{if .Issue.IsLocked}} red{{end}}" data-modal="#lock">
```

wxiaoguang commented:
I am sure there are some button colors from legacy code, like "yellow"

```
{{template "base/modal_actions_confirm" (dict "ModalButtonColors" "yellow")}}
```

wxiaoguang commented on templates/repo/code/recently_pushed_new_branches.tmpl:
```diff
@@ -5,7 +5,7 @@
 			{{ctx.Locale.Tr "repo.pulls.recently_pushed_new_branches" $branchLink $timeSince}}
 		</div>
-		<a role="button" class="ui compact positive button tw-m-0" href="{{$.Repository.ComposeBranchCompareURL $.Repository.BaseRepo .Name}}">
+		<a role="button" class="ui compact green button tw-m-0" href="{{$.Repository.ComposeBranchCompareURL $.Repository.BaseRepo .Name}}">
```
Is it really right to do so? Especially for color-blindness theme?

If we use "positive" / "negative", it is easier to define colors for color-blindness , because these colors won't be green/red.

silverwind replied:
Yes we can later refactor to success/warning/danger but I want to do this later.

https://shoelace.style/components/button#variants

wxiaoguang replied:
If it keeps "positive", the diff is truly minimal.

silverwind replied:
I can rename "green" to "success", there's only 1 instance.

silverwind replied:
Well I guess I may as well rename red now.

silverwind replied:
What do you think about success/danger instead of positive/negative?

silverwind replied:
such a rename would blow up the diff significantly because then I would also need to rename the color vars most likely. I'd rather defer this.

```css
  --color-error-border: #a04141;
  --color-error-bg: #522;
  --color-error-bg-active: #744;
  --color-error-bg-hover: #633;
  --color-error-text: #f9cbcb;
  --color-success-border: #458a57;
  --color-success-bg: #284034;
  --color-success-text: #6cc664;
  --color-warning-border: #bb9d00;
  --color-warning-bg: #3a3a30;
  --color-warning-text: #fbbd08;
  --color-info-border: #306090;
  --color-info-bg: #26354c;
  --color-info-text: #38a8e8;
```

wxiaoguang replied:
I prefer to keep a refactoring PR focused and avoid introduce unnecessary changes.

silverwind replied:
So I can do the class rename but the css vars will stay at red/green. If those approve the class names above i will do it.

wxiaoguang replied:
My preference is to just keep the "positive / negative" names. Just make them have the styles as "green / red", and leave the renaming to the future.

silverwind replied:
I prefer to not re-introduce positive/negative for now as I find success/danger to be better names. There are like 2 cases in the whole codebase that used positive/negative. green is one case, only red is much in use. So I prefer to keep red/green for now to keep the diff small and later refactor to success/danger, including color names.

wxiaoguang replied:
OK, I won't block this change.

But, "success" is a bad name to me for such button. "success" is a result, but doesn't really means what the button is for.

"danger" / "negative" either looks good to me.

silverwind replied:
Yes, success is not a ideal word but "negative" also isn't. At least these frameworks use the suggested naming scheme:

https://shoelace.style/components/button#variants
https://bulma.io/documentation/elements/button/#colors

wxiaoguang commented on templates/repo/diff/compare.tmpl:
```diff
@@ -194,9 +194,9 @@
 				<div class="four wide column middle aligned text right">
 				{{- if .PullRequest.HasMerged -}}
-				<a href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui button purple show-form">{{svg "octicon-git-merge" 16}} {{ctx.Locale.Tr "repo.pulls.view"}}</a>
+				<a href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui button primary show-form">{{svg "octicon-git-merge" 16}} {{ctx.Locale.Tr "repo.pulls.view"}}</a>
```
Sorry but it is not right.

"purple" means it has been merged.

silverwind replied:
But does it really need to indicate on the button? The button will not merge it so I think it's wrong indication.

wxiaoguang replied:
Such UI is widely used. For example, the "buttons" are provided by "label". It does provide a button-like UI.

![image](https://github.com/go-gitea/gitea/assets/2114189/c0592fbf-b8ae-4015-b5b7-3328e9fcb3a9)

silverwind replied:
Yes, but those are labels, not buttons. Nothing bad will happen if I click the red button on a merged PR, it just takes me to the PR.

silverwind replied:
So I guess I could redesign these messages to show a open/merged/closed label instead of this misleadingly colored button.

wxiaoguang replied:
> Yes, but those are labels, not buttons, e.g. they don't have actions associated.

They have, the same as this `<a>` like here.

wxiaoguang replied:
> So I guess I could redesign these messages to show a open/merged/closed label instead of this misleadingly colored button.

I could agree. Using a proper label link seems good.

wxiaoguang commented on templates/repo/diff/compare.tmpl:
```diff
@@ -194,9 +194,9 @@
+				<a href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui button primary show-form">{{svg "octicon-git-merge" 16}} {{ctx.Locale.Tr "repo.pulls.view"}}</a>
 				{{else if .Issue.IsClosed}}
-				<a href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui button red show-form">{{svg "octicon-issue-closed" 16}} {{ctx.Locale.Tr "repo.pulls.view"}}</a>
+				<a href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui button primary show-form">{{svg "octicon-issue-closed" 16}} {{ctx.Locale.Tr "repo.pulls.view"}}</a>
```
Not right again.

silverwind replied:
Same as above. The button does not do a "danger" action so I think it was wrong indication.

wxiaoguang reviewed (dismissed):
Too many nits ....

silverwind commented:
> I am sure there are some button colors from legacy code, like "yellow"
> 
> ```
> {{template "base/modal_actions_confirm" (dict "ModalButtonColors" "yellow")}}
> ```

Removed in https://github.com/go-gitea/gitea/pull/30475/commits/3b0f6d68b0ad160e7f065388c01afb6628412f4b and also fixed one issue with the "x" here:

<img width="602" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/3409b003-2ccb-4503-9836-3c86f4694925">

silverwind commented:
All nits are resolved or invalid in my opinion 😉

wxiaoguang commented on templates/repo/diff/compare.tmpl:
```diff
@@ -184,23 +184,18 @@
-						<span id="issue-title">{{RenderIssueTitle $.Context .PullRequest.Issue.Title ($.Repository.ComposeMetas ctx)}}</span>
+			<div class="existing-issue-block ui segment tw-flex tw-items-center tw-space-between tw-p-3 tw-gap-3">
+				<div class="tw-flex tw-flex-1 tw-gap-3 tw-items-center gt-ellipsis">
+					{{template "shared/issuestatelabel" (dict "." . "Issue" .PullRequest.Issue)}}
```
Please pass the dict keys/values explicitly for what the sub-template really needs.

![image](https://github.com/go-gitea/gitea/assets/2114189/2147e586-1a9e-4fa2-bf63-cdfb615ded70)

silverwind replied:
This is obsolete, it seems to work with two times `{{template "shared/issueicon" .}}` currently.

silverwind replied:
I suppose. BTW is there a difference between 

```
{{template "shared/issueicon" dict "Issue" .}}
```

and 

```
{{template "shared/issueicon" (dict "Issue" .)}}
```

I see both forms in use.

wxiaoguang replied:
IIRC no much difference actually. Just like `1+2` and `(1+2)` in other languages.

silverwind replied:
The problem is the subtemplate expects `.PullRequest` at root, so I would need to refactor all call sites...

silverwind replied:
I guess I could split out all the accessed properties...

silverwind replied:
Imho this whole `.Issue.PullRequest` and `.PullRequest.Issue` business is ripe for a major refactor in the backend where it just merges both into the same struct and differentiates via `.isPull`.

wxiaoguang replied:
If `shared/issueicon` is from legacy code, then it's fine to leave it as-is. It is out of this PR's scope.

silverwind replied:
Yeah, the untangling of issue/pr is something for later. A fix here would be 

```
{{template "shared/issueicon" dict "Issue" .PullRequest.Issue "PullRequest" .PullRequest}}
```

At all call sites, but I don't think this is worth currently without also untangling this mess in the backend.

wxiaoguang commented on templates/repo/diff/compare.tmpl:
```diff
@@ -184,23 +184,18 @@
-					<h1>
-						<span id="issue-title">{{RenderIssueTitle $.Context .PullRequest.Issue.Title ($.Repository.ComposeMetas ctx)}}</span>
+			<div class="existing-issue-block ui segment tw-flex tw-items-center tw-space-between tw-p-3 tw-gap-3">
+				<div class="tw-flex tw-flex-1 tw-gap-3 tw-items-center gt-ellipsis">
```
Why gt-ellipsis here? It doesn't work with `flex` (IIRC we have discussed it in another PR).

I think this `gt-ellipsis` could be removed, because the children elements could have proper gt-ellipsis

silverwind replied:
What is the proper fix? I found using `gt-ellipsis` two times works reliable.

wxiaoguang replied:
There is already `<div class="issue-title gt-ellipsis">` below, why this "flex" needs `gt-ellipsis`? Flex doesn't work with it ....

silverwind replied:
`min-width: 0` on the parent works, but I still need `gt-ellipsis` in the child to be ellipsed.

silverwind commented:
@wxiaoguang sorry, last commit was not ready yet for review and I force-pushed over it. It's now simply using `issueicon` template two times:

<img width="1333" alt="Screenshot 2024-04-14 at 18 09 11" src="https://github.com/go-gitea/gitea/assets/115237/735ac855-372e-4575-944b-b73ea4a41e68">

wxiaoguang commented:
> @wxiaoguang sorry, last commit was not ready yet for review and I force-pushed over it so [8229c2d](https://github.com/go-gitea/gitea/commit/8229c2d16bfd3c1db41983884e1412fc459ed7c2) has all the changes. It's now simply using `issueicon` template two times:

I think my reviews still apply.

* Use explicit keys instead of "." for sub-templates
* Remove unnecessary `gt-ellipsis`

silverwind commented:
> Use explicit keys instead of "." for sub-templates

This would affect at least 4-5 more places and I'm not really willing to further blow this PR out of scope.

wxiaoguang commented:
`use tw-min-w-0 on parent` that's not right.

See my example: https://github.com/go-gitea/gitea/pull/30344#discussion_r1564455835

wxiaoguang commented on templates/repo/diff/compare.tmpl:
```diff
@@ -192,7 +192,7 @@
 					</div>
 				</div>
-				<a href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui compact button primary">
+				<a role="button" href="{{$.RepoLink}}/pulls/{{.PullRequest.Issue.Index}}" class="ui compact button primary">
```
I believe it is a link but not a button.

A button means it do something on current page, a link means it navigates to another place.

Here, it navigates, so it can't be a button.

silverwind replied:
It renders like a button, but whatever, I can revert.

silverwind commented:
> `use tw-min-w-0 on parent` that's not right.
> 
> See my example: [#30344 (comment)](https://github.com/go-gitea/gitea/pull/30344#discussion_r1564455835)

Your example does not work here:

<img width="1346" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/e465fdf4-793c-432f-b6b6-8db0093dcc09">

wxiaoguang commented:
> > `use tw-min-w-0 on parent` that's not right.
> > See my example: [#30344 (comment)](https://github.com/go-gitea/gitea/pull/30344#discussion_r1564455835)
> 
> Your example does not work here. I think properties need to be set on at least two nodes:

My example should be right.

Here it doesn't work because you have "flex" on `issue-title`

silverwind commented:
> > > `use tw-min-w-0 on parent` that's not right.
> > > See my example: [#30344 (comment)](https://github.com/go-gitea/gitea/pull/30344#discussion_r1564455835)
> > 
> > 
> > Your example does not work here. I think properties need to be set on at least two nodes:
> 
> My example should be right.
> 
> Here it doesn't work because you have "flex" on `issue-title`

Check it out yourself, I see no solution that fixes the ellipsis by adjusting gt-ellipsis only.

silverwind commented:
To explain: This is the flex child where min-width:0 works:

<img width="1333" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/42a6901d-a154-4220-8761-873252f5e5d0">

This is the grand-child where it does not work:

<img width="1331" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/8ac9abf7-fd1c-453b-b3e2-1d75ec523062">

So it's a nested flexbox and ellipsis only works on top-level apparently.

wxiaoguang commented:
> Check it out yourself, I see no solution that fixes the ellipsis by adjusting the element with gt-ellipsis only.

I see the problem. There are 2 different "issue-title", one is here, another is for ".view.issue", that's really annoying. Will try to fix

silverwind commented:
Could also just make it wrap here. There's no real reason to ellipsis this with som much space on the page.

wxiaoguang commented:
![image](https://github.com/go-gitea/gitea/assets/2114189/098ee426-4345-4dd1-b081-b130e7f7f1d2)

![image](https://github.com/go-gitea/gitea/assets/2114189/d9828a29-86de-4287-a0ac-bb16473e8937)

silverwind commented:
Hmm I see no effect when I disable the `min-width: 0` rule. Are you sure it's essential?

wxiaoguang commented:
> Hmm I see no effect when I disable the `min-width: 0` rule. Are you sure it's essential?

For this case, it have no effect because its parent is not flex. But I believe it is right to add min-width for it. There are many references in https://github.com/go-gitea/gitea/pull/30344#discussion_r1564455835

silverwind commented:
Also, I would like some `gap`, but that's not possible without flexbox which would break the truncation again. Anyways. I will now make it wrap I guess.

wxiaoguang commented:
> Also, I would like some `gap`, but that's not possible without flexbox which would break the truncation again.

According to the screenshot, I don't see any "gap" is needed for this case. Feel free to wrap it.

silverwind commented:
Wrap done. Also added a upcoming tailwind class.

<img width="1332" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/3bd4b1cd-238c-490d-8d51-9a590e6bd4e3">

silverwind commented:
Increased the gap. I think it looks better because the spacing between icon and content is similar to the box above:

<img width="1334" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/cd3590f1-3cba-40b9-92a8-28d5eb357909">

`flex-text-block` already has a small gap, but this needed more. I think we are abusing that class a bit here, but whatever.

wxiaoguang commented:
> `flex-text-block` already has a small gap, but this needed more. I think we are abusing that class a bit here, but whatever.

IMO it was designed to work for this case: if default gap doesn't look good, add some `tw` helpers to fine tune. `flex-xxx` only provides basic flex ability to align items.

silverwind commented:
One more tweak added. All buttons will now not wrap their text until mobile viewport. I think it is worth to try this because wrapped text in buttons almost always looks bad in desktop view.

<img width="193" alt="Screenshot 2024-04-14 at 19 05 07" src="https://github.com/go-gitea/gitea/assets/115237/dc9c640c-65e2-438d-9664-6e4c455ab5f5">

<img width="602" alt="Screenshot 2024-04-14 at 19 05 00" src="https://github.com/go-gitea/gitea/assets/115237/2c069c55-221e-41f6-be3c-1524cef665c3">

wxiaoguang commented on web_src/css/helpers.css:
```diff
@@ -52,6 +52,9 @@ only use:
 .tw-hidden.tw-hidden { display: none !important; }
 
+/* proposed class from https://github.com/tailwindlabs/tailwindcss/pull/12128 */
+.tw-break-anywhere { overflow-wrap: anywhere; }
```
```suggestion
.tw-break-anywhere { overflow-wrap: anywhere !important; }
```

to match others.
